### ISSUE TITLE
Support for multiple subnet IDs

### ISSUE THREAD

**Issue by fermanjj** [RaJiska/terraform-aws-fck-nat#38]
Currently it appears the module supports only a single subnet ID. This means if you're using spot instances in an ASG, you can get denied provisioning an instance due to limited capacity in that specific Availability zone you've allocated the subnet to.

The exact error for more context:

```
We currently do not have sufficient t4g.nano capacity in the Availability Zone you requested (us-east-1a). Our system will be working on provisioning additional capacity. You can currently get t4g.nano capacity by not specifying an Availability Zone in your request or choosing us-east-1b, us-east-1c, us-east-1d, us-east-1f. Launching EC2 instance failed.
```

**Comment by RaJiska** [comment]
Just as for AWS' NAT Gateway service, this module only spans across one availability zone. You might want to use multiple instance types within your ASG to mitigate the issue of instance shortage when using spot instances

**Comment by callumthomson** [comment]
@RaJiska 

> Just as for AWS' NAT Gateway service, this module only spans across one availability zone.

I would agree, but this is a problem unique to spot instances, which does not apply to NAT gateways and I think this module can solve it easily.

While launching an EC2 instance requires a subnet, In HA mode specifically, the ASG can and should be created to include multiple subnets. I had this error today on my fck-nat instance that was already running and was interupted:

`
Launching a new EC2 instance. Status Reason: We currently do not have sufficient t4g.nano capacity in the Availability Zone you requested (eu-west-2a). Our system will be working on provisioning additional capacity. You can currently get t4g.nano capacity by not specifying an Availability Zone in your request or choosing eu-west-2b, eu-west-2c. Launching EC2 instance failed.
`

Because of the way this module creates the ASG, it was not able to replace the instance in another subnet.

> You might want to use multiple instance types within your ASG to mitigate the issue of instance shortage when using spot instances

It doesn't seem like there is an option to specify more than 1 instance type for the ASG in this module. The only other option that I can think of would be to create multiple instances of this module with a different instance type/subnet id, therefore creating multiple NAT instances. This means higher cost.

I think multiple subnets should be an option when using an ASG (HA mode). If you want to limit to a single subnet, then you should also have that option.

**Comment by asmforce** [comment]
@RaJiska I thought the sole purpose of HA mode was to be able to use ASG to quickly recover from an AZ failure which arguably is the main possible cause for NAT EC2 instance to fail. So I was quite surprised to figure out it was not.

Is there any reason NOT to support multiple subnets (hence multiple AZs) per ASG in HA mode?

**Comment by RaJiska** [comment]
> I thought the sole purpose of HA mode was to be able to use ASG to quickly recover from an AZ failure which arguably is the main possible cause for NAT EC2 instance to fail. So I was quite surprised to figure out it was not.

This was already addressed above, you can have Multi-AZ, just the same way you have it with NAT Gateway, you deploy one instance of the module per AZ, and you ensure your workload spans those AZs.

> Is there any reason NOT to support multiple subnets (hence multiple AZs) per ASG in HA mode?

This is because routing is done at the subnet level and points to an ENI which spans a single subnet at most.

**Comment by RaJiska** [comment]
@callumthomson Apologies for the late response. I'll still give an answer in case this helps others.

It feels like we are mixing two different things here:
1. Reliability failure due to depleted spot pool over one instance type
2. Multi-AZ / subnet for reliability

The first one is indeed a problem, which can be partially addressed by adding the option to support multiple instances types on the ASG.

The second is already possible by instantiating the module over multiple AZs. Unfortunately it won't be possible to have a single module spanning multiple AZs due to the technical constraints mentioned above.
